### PR TITLE
[Feature] Introduce a new framework that supports scheduling for Formats and Processes

### PR DESCRIPTION
**Pull request by LiangDai-Mars**
## Why are the changes needed?
 
To introduce a new framework that supports scheduling for a wider range of Formats and Processes

## Brief change log
 
- Introduce a new top-level service, ProcessService, responsible for managing the lifecycle of Processes.
- Introduce a new plugin type **ActionCoordinator** to extend how TableProcess is created.
- Introduce a new plugin type **ExecutionEngine** to extend how external **Process** is executed.

## How was this patch tested?

- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinator.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  String executionEngine();
+
+  long getNextExecutingTime(TableRuntime tableRuntime);
+
+  boolean enabled(TableRuntime tableRuntime);
+
+  long getExecutorDelay();
+
+  TableProcess<TableProcessState> getTableProcess(TableRuntime tableRuntime);
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinator.java]: The active process is managed by ProcessService or TableRuntime.

ActionCoordinator should be  stateless.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    }
+  }
+
+  protected void cancel(TableRuntime tableRuntime, TableProcess<TableProcessState> process) {
+    process = coordinator.cancelTableProcess(tableRuntime, process);
+    processService.cancel(process);
+  }
+
+  protected boolean hasAliveTableProcess(TableRuntime tableRuntime) {
+    TableProcess<TableProcessState> process = coordinator.getTableProcess(tableRuntime);
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: Should not fetch active process from coordinator.

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  void dispose();
+
+  void synchronizedInvoke(Runnable runnable);
+
+  TableProcessOperation begin();
+
+  interface TableProcessOperation {
+
+    TableProcessOperation insertTableProcess(TableProcessState state);
```
Comment (baiyangtx) [inline amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java]: NO INSERT.

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    TableProcessOperation updateTableProcess(TableProcessState state);
```
Comment (baiyangtx) [inline amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java]: Refer to TableRuntimeStore.

Comment (baiyangtx) [reply amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java]: Updatable fields.
- Parameters

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcess.java [hunk 5]
```diff
@@ -29,18 +29,30 @@ public abstract class TableProcess<T extends TableProcessState> implements Amoro
 
   protected final T state;
   protected final TableRuntime tableRuntime;
+  protected final TableProcessStore tableProcessStore;
```
Comment (baiyangtx) [inline amoro-common/src/main/java/org/apache/amoro/process/TableProcess.java:31]: Remove `T state` from TableProcess.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/executor/ExecuteEngine.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  /**
+   * Submit a table process to engine.
+   *
+   * <p>This method must returned when process status is not Pending.
+   */
+  String submitTableProcess(TableProcess<TableProcessState> tableProcess);
+
+  /** Cancel a table process in engine. */
+  ProcessStatus cancelTableProcess(
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/executor/ExecuteEngine.java]: void tryCancel(String processIdentifier);

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        new TableProcessExecutor(process, executeEngine, ExecuteOption.RUN);
+    executor.onProcessFinished(
+        () -> {
+          ActionCoordinatorScheduler scheduler =
+              actionCoordinators.get(process.getState().getAction().getName());
+          if (scheduler != null
+              && process.getStatus() == ProcessStatus.FAILED
+              && process.getState().getRetryNumber() < scheduler.PROCESS_MAX_RETRY_NUMBER
+              && process.getTableRuntime() != null) {
+            process.getState().addRetryNumber();
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java]: use TableProcessStore.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+  @VisibleForTesting
+  public TableProcessTracker getTableProcessTracker() {
+    return tableProcessTracker;
+  }
+
+  public static class TableProcessTracker {
+    private final Map<ServerTableIdentifier, TableProcess> activeTableProcess =
+        new ConcurrentHashMap<>();
+    private Map<String, String> properties;
+
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java]: Map<ProcessId, TableProcess> 
Map<TableIdentifier, Set<ProcessId>>.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+  @Override
+  public void handleTableRemoved(TableRuntime tableRuntime) {
+    TableProcess process =
+        processService
+            .getTableProcessTracker()
+            .getTableProcessInstance(tableRuntime.getTableIdentifier());
+    if (process != null) {
+      cancel(tableRuntime, process);
+    }
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: ProcessService could do this.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+              .getTableProcessTracker()
+              .getTableProcessInstance(tableRuntime.getTableIdentifier());
+      LOG.warn(
+          "Detect table process: {} with status: {} exists for table runtime: {}, skip recover {} action this time.",
+          process.getId(),
+          process.getStatus(),
+          tableRuntime.getTableIdentifier(),
+          getAction());
+    } else {
+      TableProcess process = coordinator.recoverTableProcess(tableRuntime, processMeta);
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: processStore

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+          tableRuntime.getTableIdentifier(),
+          getAction());
+    } else {
+      TableProcess process = coordinator.createTableProcess(tableRuntime);
+      processService.register(tableRuntime, process);
+    }
+  }
+
+  protected void recover(TableRuntime tableRuntime, TableProcessMeta processMeta) {
+    if (hasAliveTableProcess(tableRuntime)) {
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: All processMeta must be recovered.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+            .getTableProcessInstance(tableRuntime.getTableIdentifier());
+    if (existProcess != null && existProcess.getId() == process.getId()) {
+      process = coordinator.retryTableProcess(process);
+      processService.retry(process);
+    } else {
+      LOG.warn(
+          "Detect no table process exists or exist table process id is different from retry process for table runtime: {}, skip retry {} action this time.",
+          tableRuntime.getTableIdentifier(),
+          getAction());
+    }
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: No need to check here.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+    }
+  }
+
+  public void recover(TableRuntime tableRuntime, TableProcess process) {
+    synchronized (process) {
+      // TODO: init some status
+      tableProcessTracker.trackTableProcess(tableRuntime.getTableIdentifier(), process);
+      executeOrTraceProcess(process);
+    }
+  }
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java:127]: private method. same with below.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/executor/TableProcessExecutor.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+      if (t instanceof InterruptedException) {
+        LOG.info(
+            "Table process {} with identifier {} may have been interrupted by canceling process, exit submit process.",
+            tableProcess.getId(),
+            externalProcessIdentifier);
+        return;
+      }
+      status = ProcessStatus.FAILED;
+      message = t.getMessage();
+    }
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/executor/TableProcessExecutor.java:120]: The thread **must** set process status to a finnal status before it return.

File: amoro-ams/src/test/java/org/apache/amoro/server/process/TestExecuteEngine.java [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+
+    @Override
+    public void run() {
+      // 1. Generate a random integer between 1 and 100 (recommend ThreadLocalRandom for
+      // thread-safety and high performance)
+      int randomNum = ThreadLocalRandom.current().nextInt(1, 101);
+      System.out.println("Generated random number: " + randomNum);
+      System.out.println("Start loop: i from 0 to " + randomNum + ", sleep 5 seconds each time");
+      System.out.println("----------------------------------------");
+
```
Comment (baiyangtx) [inline amoro-ams/src/test/java/org/apache/amoro/server/process/TestExecuteEngine.java]: Remove println, replace them with LOG.info

File: amoro-ams/src/test/java/org/apache/amoro/server/process/TestExecuteEngine.java [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+import java.security.SecureRandom;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class TestExecuteEngine implements ExecuteEngine {
```
Comment (baiyangtx) [inline amoro-ams/src/test/java/org/apache/amoro/server/process/TestExecuteEngine.java]: Maybe MockExecuteEngine is better.

File: amoro-ams/src/test/java/org/apache/amoro/server/process/MockActionCoordinator.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.amoro.shade.guava32.com.google.common.annotations.VisibleForTesting;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TestActionCoordinator implements ActionCoordinator {
+  public static final int PROCESS_MAX_POOL_SIZE = 1000;
+  private static final TableFormat[] DEFAULT_FORMATS = new TableFormat[] {TableFormat.PAIMON};
+
+  public static final Action DEFAULT_ACTION = new Action(DEFAULT_FORMATS, 0, "default_action");
```
Comment (baiyangtx) [inline amoro-ams/src/test/java/org/apache/amoro/server/process/MockActionCoordinator.java:35]: MOCK_ACTION maybe better.

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcess.java (hunk 5 above)
Comment (LiangDai-Mars) [reply amoro-common/src/main/java/org/apache/amoro/process/TableProcess.java]: done

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinator.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+package org.apache.amoro.server.process;
+
+import org.apache.amoro.Action;
+import org.apache.amoro.ActivePlugin;
+import org.apache.amoro.TableFormat;
+import org.apache.amoro.TableRuntime;
+import org.apache.amoro.process.TableProcess;
+import org.apache.amoro.process.TableProcessStore;
+import org.apache.amoro.server.utils.SnowflakeIdGenerator;
+
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinator.java:28]: class docs and method docs

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+import org.apache.amoro.TableFormat;
+import org.apache.amoro.TableRuntime;
+import org.apache.amoro.process.TableProcess;
+import org.apache.amoro.process.TableProcessStore;
+import org.apache.amoro.server.scheduler.PeriodicTableScheduler;
+import org.apache.amoro.server.table.TableService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java:29]: class docs

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+    this.actionCoordinatorManager = actionCoordinatorManager;
+    this.executeEngineManager = executeEngineManager;
+  }
+
+  public RuntimeHandlerChain getTableHandlerChain() {
+    return tableRuntimeHandler;
+  }
+
+  public void register(TableRuntime tableRuntime, TableProcess process) {
+    synchronized (process) {
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java]: Synchronized by process maybe not work.

File: amoro-ams/src/main/java/org/apache/amoro/server/process/executor/EngineType.java [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.amoro.server.process.executor;
+
+public enum EngineType {
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/executor/EngineType.java]: EngineType should be an extensible enumeration type, like `Action`

File: amoro-ams/src/main/java/org/apache/amoro/server/process/executor/ExecuteOption.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.amoro.server.process.executor;
+
+public enum ExecuteOption {
```
Comment (baiyangtx) [inline amoro-ams/src/main/java/org/apache/amoro/server/process/executor/ExecuteOption.java]: This class seems useless.

File: amoro-ams/src/main/resources/mysql/upgrade.sql [hunk 23]
```diff
@@ -155,3 +155,9 @@ DROP TABLE IF EXISTS table_runtime_old;
 
 -- ADD bucket_id to table_runtime
 ALTER TABLE `table_runtime` ADD COLUMN `bucket_id` VARCHAR(4)  DEFAULT NULL COMMENT 'Bucket number to which the record table belongs';
+
+-- ADD external_process_identifier, retry_number, process_parameters to table_runtime
+ALTER TABLE `table_process`
+ADD COLUMN `external_process_identifier` varchar(256) DEFAULT NULL COMMENT 'Table optimizing external process identifier',
+ADD COLUMN `retry_number` int(11) NOT NULL COMMENT 'Retry times',
```
Comment (baiyangtx) [inline amoro-ams/src/main/resources/mysql/upgrade.sql]: `NOT NULL default 0`

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcess.java [hunk 24]
```diff
@@ -19,36 +19,237 @@
...
+            && newStatus == ProcessStatus.KILLED;
+      case RETRY_REQUESTED:
+        return !terminalRequested
+            && (getStatus() == ProcessStatus.FAILED && newStatus == ProcessStatus.PENDING);
+      default:
+        return false;
+    }
+  }
+
+  public boolean tryTransitState(
```
Comment (baiyangtx) [inline amoro-common/src/main/java/org/apache/amoro/process/TableProcess.java]: I think the `transitState` method should be in `TableProcessStore`.

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.amoro.process;
+
+import org.apache.amoro.Action;
+
+import java.util.Map;
+
+public interface TableProcessStore {
```
Comment (baiyangtx) [inline amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java:26]: docs

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinator.java (hunk 1 above)
Comment (LiangDai-Mars) [reply amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinator.java]: done

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java (hunk 3 above)
Comment (LiangDai-Mars) [reply amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java]: done

File: amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java (hunk 4 above)
Comment (LiangDai-Mars) [reply amoro-common/src/main/java/org/apache/amoro/process/TableProcessStore.java]: done

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java (hunk 10 above)
Comment (LiangDai-Mars) [reply amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: done

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java (hunk 12 above)
Comment (LiangDai-Mars) [reply amoro-ams/src/main/java/org/apache/amoro/server/process/ActionCoordinatorScheduler.java]: done

File: amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java (hunk 13 above)
Comment (LiangDai-Mars) [reply amoro-ams/src/main/java/org/apache/amoro/server/process/ProcessService.java]: done

File: amoro-ams/src/test/java/org/apache/amoro/server/process/TestExecuteEngine.java (hunk 15 above)
Comment (LiangDai-Mars) [reply amoro-ams/src/test/java/org/apache/amoro/server/process/TestExecuteEngine.java]: done

File: amoro-ams/src/main/resources/mysql/upgrade.sql (hunk 23 above)
Comment (LiangDai-Mars) [reply amoro-ams/src/main/resources/mysql/upgrade.sql]: done
